lgc/patch/PatchResourceCollect.cpp

```diff
@@ -283,6 +260,130 @@ void PatchResourceCollect::setNggControl() {
+  const auto polygonMode = m_pipelineState->getRasterizerState().polygonMode;
+  if (polygonMode == PolygonModeLine || polygonMode == PolygonModePoint) {
+    return false;
+  }
+
+  // Check the presence of position export, disable NGG culling if position export is absent
+  bool usePosition = false;
+  if (hasGs)
+    usePosition = m_pipelineState->getShaderResourceUsage(ShaderStageGeometry)->builtInUsage.gs.position;
+  else if (hasTs)
+    usePosition = m_pipelineState->getShaderResourceUsage(ShaderStageTessEval)->builtInUsage.tes.position;
+  else
+    usePosition = m_pipelineState->getShaderResourceUsage(ShaderStageVertex)->builtInUsage.vs.position;
+
+  if (!usePosition)
+    return false; // No position export
+
+  // Find position export call
+  std::string posCallName = lgcName::OutputExportBuiltIn;
+  posCallName += PipelineState::getBuiltInName(BuiltInPosition);
+  auto callStage = hasGs ? ShaderStageGeometry : (hasTs ? ShaderStageTessEval : ShaderStageVertex);
+
+  CallInst *posCall = nullptr;
+  for (Function &func : *module) {
+    if (func.getName().startswith(posCallName)) {
+      for (User *user : func.users()) {
+        auto call = cast<CallInst>(user);
+        if (m_pipelineShaders->getShaderStage(call->getFunction()) == callStage) {
+          posCall = call;
+          break;
+        }
+      }
+
+      if (posCall) // Already find the call
+        break;
+    }
+  }
+  assert(posCall); // Position export must exist
+
+  // Check position value, disable NGG culling if it is produced by instructions
+  auto posValue = posCall->getArgOperand(posCall->getNumArgOperands() - 1); // Last argument is position value
+  if (!isa<Instruction>(posValue))
+    return false;
+
+  // For GS, we can safely enable NGG culling here; for non-GS, still need further analysis
+  if (hasGs)
+    return true;
+
+  // Check if the sources of position value are eventually produced by vertex-based inputs loaded from memory
+  SmallVector<Instruction *, 32> posSources;
+  posSources.push_back(cast<Instruction>(posValue));
+
+  while (!posSources.empty()) {
+    auto posSource = posSources.back();
+    posSources.pop_back();
+
+    // Pop up a source and check its operands
+    for (auto &operand : posSource->operands()) {
+      if (isa<Instruction>(operand)) {
+        // Only check instructions
+        auto inst = cast<Instruction>(operand);
+        if (isa<CallInst>(inst)) {
+          // Further check calls
+          auto calleeName = cast<CallInst>(inst)->getCalledFunction()->getName();
+          // Generic inputs are always vertex-based
+          if (calleeName.startswith(lgcName::InputImportGeneric))
+            return true;
+          // If tessellation, we have to further check certain built-in inputs that are vertex-based
+          if (hasTs && calleeName.startswith(lgcName::InputImportBuiltIn)) {
+            auto builtInName = calleeName.slice(strlen(lgcName::InputImportBuiltIn), StringRef::npos);
```

Comment:
nit: `calleeName.drop(strlen(lgcName::InputImportBuiltIn)`

Reply:
Thanks. The change is fine.